Remove the need for a run-function in the python code

lisajulia opened the pull request:
- Make the EclipseState, Schedule, SummaryState available as attributes of the module opm_embedded instead
- This enables tooltips for writing code in a python ide by typing opm_embedded.<tab>
- The code will still be backwards compatible

lisajulia commented:
@blattms @hakonhagland: can you have a look here?
I'll still add a test to verify the backwards compatibility and then I still need to add the path of the opm_embedded library to the systems PYTHONPATH.

lisajulia commented:
jenkins build this please

blattms commented:
jenkins build this please

lisajulia commented:
jenkins build this please

lisajulia commented:
jenkins build this please

hakonhagland commented on opm/input/eclipse/Python/PyRunModule.cpp:
```diff
@@ -82,10 +67,84 @@ py::cpp_function py_actionx_callback(const std::function<void(const std::string&
+                    
+import opm_embedded
+
+help(opm_embedded.current_ecl_state)
```
Yes this works in IPython3 :smile:, just remember to select the correct Python interpreter that was used to compile opm_embbeded with. However, in VS Code intellisense and autocompletion does not work for some reason

blattms commented:
Currently, opm_embedded is also built if `OPM_ENABLE_EMBEDDED_PYTHON` is `OFF`. Shouldn't we change that?

blattms commented on python/cxx/export.cpp:
```diff
@@ -28,3 +31,40 @@ void python::common::export_all(py::module& module) {
+        -----------------------
+
+        .. currentmodule:: opm_embedded
+
```
We should probably mention the 4 attributes above here, if that is possible.

blattms commented on python/cxx/export.cpp:
```diff
@@ -28,3 +31,40 @@ void python::common::export_all(py::module& module) {
+    module.attr("current_ecl_state") = std::make_shared<EclipseState>();
+    module.attr("current_summary_state") = std::make_shared<SummaryState>();
+    module.attr("current_schedule") = std::make_shared<Schedule>();
+    module.attr("report_step") = 0;
```
All the other attributess are prefixed with "current_"...

lisajulia commented:
> Currently, opm_embedded is also built if `OPM_ENABLE_EMBEDDED_PYTHON` is `OFF`. Shouldn't we change that?

Thanks! I've changed that now.

lisajulia commented:
jenkins build this please

lisajulia commented:
jenkins build this please

blattms commented:
Sorry, I merged your other PR and this needs a rebase now.
We should also install opm_embedded, see lisajulia/opm-common#1

lisajulia commented:
jenkins build this please

lisajulia commented:
jenkins build this please

lisajulia commented:
> Sorry, I merged your other PR and this needs a rebase now. We should also install opm_embedded, see [lisajulia#1](https://github.com/lisajulia/opm-common/pull/1)

I've added your commit and rebased this branch, so it has no conflicts anymore.

blattms commented:
Perfect. Seems like some tests are failing. Maybe you renamed the report_step attribute and did not change all scripts?

lisajulia commented:
jenkins build this please

blattms commented on opm/input/eclipse/Python/PyRunModule.cpp:
```diff
@@ -45,33 +45,18 @@ PyRunModule::PyRunModule(std::shared_ptr<const Python> python, const std::string
+    try {
+        this->opm_embedded = py::module::import("opm_embedded");
+    } catch (const std::exception& e) {
+        std::cerr << "Exception thrown when loading Python module opm_embedded: " << e.what() << std::endl;
+        throw e;
+    } catch (...) {
+        std::cerr << "General exception thrown when loading Python module opm_embedded!" << std::endl;
```
If this is catched and handled somewhere else, then it might be better to create a new exception with a better error message that indicates that this happend when running PYACTION.

If that is exception is handled elsewhere and logged we do not need any message here. If not we should use something like
```
OpmLog::error(fmt::format("Exception thrown when loading Python module opm_embedded: {}), e.what());
```

blattms commented on opm/input/eclipse/Python/PyRunModule.cpp:
```diff
@@ -82,10 +67,84 @@ py::cpp_function py_actionx_callback(const std::function<void(const std::string&
+            }
+            this->module = py::module::import(this->module_name.c_str());
+        } catch (const std::exception& e) {
+            std::cerr << "Exception thrown when loading Python module " << this->module_name << ": " << e.what() << std::endl;
+            throw e;
+        } catch (...) {
+            std::cerr << "General exception thrown when loading Python module " << this->module_name << "!" << std::endl;
```
here, too

blattms commented on opm/input/eclipse/Python/PyRunModule.cpp:
```diff
@@ -82,10 +67,84 @@ py::cpp_function py_actionx_callback(const std::function<void(const std::string&
+        this->module.attr("storage") = this->storage;
+
+        if (py::hasattr(this->module, "run")) {
+            std::cout << R"(PYACTION can be used without a run(ecl_state, schedule, report_step, summary_state, actionx_callback) function, its arguments are available as attributes of the module opm_embedded, try the following in your python script:
```
I think we should use `OpmLog::info` here. Then the message is also in the output files.

blattms reviewed (commented):
This looks really good. Sorry that I did not spot the output to std::cout and std::cerr previously.

I think we should use the logging mechanism for that as it will make sure that the message are also in the files. The question here is whether the rethrown exceptions are already catched and logged somewhere else. In that case we do  not even need to log them here.

One question:
What happens if there is a PYACTION, but flow was built without embedded python? The simulator will abort, right?

lisajulia commented:
> This looks really good. Sorry that I did not spot the output to std::cout and std::cerr previously.
> 
> I think we should use the logging mechanism for that as it will make sure that the message are also in the files. The question here is whether the rethrown exceptions are already catched and logged somewhere else. In that case we do not even need to log them here.
> 
> One question: What happens if there is a PYACTION, but flow was built without embedded python? The simulator will abort, right?

For the last question, see: https://github.com/OPM/opm-common/pull/3998

lisajulia commented:
jenkins build this please

blattms reviewed (approved):
Looks very good to me. Just one minor nitpick concerning an include.

lisajulia commented:
jenkins build this please

blattms commented:
What import issue?

blattms commented on opm/input/eclipse/Python/PyRunModule.cpp:
```diff
@@ -21,13 +21,16 @@ error BUG: The PyRunModule.hpp header should *not* be included in a configuratio
 #include "PyRunModule.hpp"
 
-#include <filesystem>
+#include <iostream>
```
This should not be needed.

(I thought I made that comment yesterday. Apparently, not. Maybe forget to press the right burron.)

blattms replied:
```suggestion
```

lisajulia commented:
> What import issue?

Ok, then I think I've resolved all issues :). Yet, for some reason I can't start Jenkins here.. @blattms can you do that and then merge?

blattms commented:
jenkins build this please

blattms commented:
Cool. Thanks a lot.
